Upgrade ASLayoutElementContext to an Object #trivial

Adlai-Holler opened the pull request:
This gives us a ton more flexibility for the future.

I'm on vacation so this was a little bit thrown-together, but I'm excited about it! Please give it a thorough review.

The GitHub diff got a little confused, so some lines that didn't change at all appear to have changed, like:

```
CGFloat const ASLayoutElementParentDimensionUndefined = NAN;
CGSize const ASLayoutElementParentSizeUndefined = {ASLayoutElementParentDimensionUndefined, ASLayoutElementParentDimensionUndefined};
```

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: Any source code changes should...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Adlai-Holler commented:
Note:  changed the names to push/pop because the semantics around what operations are allowed weren't totally clear once I moved to an object. Also, of course it would be cleaner to have a separate file and maybe a custom initializer, or even a class-method interface to this stuff, but I just wanted to get the basic approach up.

nguyenhuy commented on Source/Layout/ASLayoutElement.mm:
```diff
@@ -30,69 +30,63 @@
 
+static void ASLayoutElementDestructor(void *p) {
+  if (p != NULL) {
+  	ASDisplayNodeCFailAssert(@"Thread exited without clearing layout element context!");
```
Nit: Indentation looks off.

nguyenhuy commented on Source/Layout/ASLayoutElement.mm:
```diff
@@ -30,69 +30,63 @@
   static dispatch_once_t onceToken;
   dispatch_once(&onceToken, ^{
-    pthread_key_create(&ASLayoutElementContextKey, free);
+    // NOTE: NULL destructorm means that if a thread
```
What's up?

Adlai-Holler commented:
Hey Huy, thanks for the review
